packages/safe-ds-vscode/src/extension/pythonServer.ts

```diff
@@ -0,0 +1,334 @@
+ *
+ * @param executionId Id that uniquely identifies the execution that produced this stack frame
+ * @param frame Stack frame from the python execution
+ */
+export const tryMapToSafeDSSource = async function (
+    executionId: string,
+    frame: RuntimeErrorBacktraceFrame | undefined,
+): Promise<RuntimeErrorBacktraceFrame | undefined> {
+    if (!frame) {
+        return undefined;
+    }
+    if (!executionInformation.has(executionId)) {
+        return undefined;
+    }
+    const execInfo = executionInformation.get(executionId)!;
+    let sourceMapKeys = Array.from(execInfo.generatedSource.keys() || []).filter((value) =>
+        value.endsWith(`${frame.file}.py.map`),
+    );
+    if (sourceMapKeys.length === 0) {
+        return undefined;
+    }
+    let sourceMapKey = sourceMapKeys[0]!;
+    if (!execInfo.sourceMappings.has(sourceMapKey)) {
+        const sourceMapObject = JSON.parse(execInfo.generatedSource.get(sourceMapKey)!);
+        sourceMapObject.sourcesContent = [execInfo.source];
+        const consumer = await new SourceMapConsumer(sourceMapObject);
+        execInfo.sourceMappings.set(sourceMapKey, consumer);
+    }
+    const outputPosition = execInfo.sourceMappings.get(sourceMapKey)!.originalPositionFor({
+        line: Number(frame.line),
+        column: 0,
+        bias: SourceMapConsumer.LEAST_UPPER_BOUND,
+    });
+    return { file: outputPosition.source || '<unknown>', line: outputPosition.line || 0 };
+};
+
+/**
+ * Execute a Safe-DS pipeline on the python runner.
+ *
+ * @param services SafeDsServices object, used to import the pipeline file.
+ * @param pipelinePath Path to a Safe-DS pipeline file to execute.
+ * @param id A unique id that is used in further communication with this pipeline.
+ */
+export const executePipeline = async function (services: SafeDsServices, pipelinePath: string, id: string) {
+    // TODO include all relevant files from workspace
+    const documentUri = URI.file(pipelinePath);
+    const workspaceRoot = path.dirname(documentUri.fsPath); // TODO get actual workspace root
+    services.shared.workspace.LangiumDocuments.deleteDocument(documentUri);
+    let document;
+    try {
+        document = await importPipeline(services, documentUri);
+    } catch (e) {
+        if (e instanceof Error) {
+            logError(e.message);
+            vscode.window.showErrorMessage(e.message);
+        }
+        return;
+    }
+    const lastExecutedSource = document.textDocument.getText();
+
+    const node = document.parseResult.value;
+    //
+    let mainPipelineName;
+    let mainModuleName;
+    if (!ast.isSdsModule(node)) {
+        return;
+    }
+    const mainPythonModuleName = services.builtins.Annotations.getPythonModule(node);
+    const mainPackage = mainPythonModuleName === undefined ? node?.name.split('.') : [mainPythonModuleName];
+    const pipelines = (node?.members?.filter(ast.isSdsModuleMember) ?? []).filter(ast.isSdsPipeline);
```

Comment:
```suggestion
    const pipelines = getModuleMembers(node).filter(ast.isSdsPipeline);
```

An import must be added as well for `getModuleMembers`.

Reply:
Also replace the `filter` with `find`, which is faster and better conveys your intention.